File: backend/src/main/java/com/finance/dto/CategoryDtos.java
package com.finance.dto;

import com.finance.entities.CategoryType;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import java.util.UUID;

public class CategoryDtos {

    public record CategoryRequest(
            @NotBlank String name,
            @NotNull CategoryType type,
            @NotBlank String color,
            @NotBlank String icon
    ) {
    }

    public record CategoryResponse(
            UUID id,
            String name,
            CategoryType type,
            String color,
            String icon,
            boolean archived
    ) {
    }
}

File: backend-dotnet/FinanceTracker.Api/Contracts/CategoryDtos.cs
using System.ComponentModel.DataAnnotations;
using FinanceTracker.Api.Entities;
using FinanceTracker.Api.Infrastructure;

namespace FinanceTracker.Api.Contracts;

public sealed record CategoryRequest(
    [param: NotBlank(ErrorMessage = "must not be blank")]
    string? name,
    [param: Required(ErrorMessage = "must not be null")]
    CategoryType? type,
    [param: NotBlank(ErrorMessage = "must not be blank")]
    string? color,
    [param: NotBlank(ErrorMessage = "must not be blank")]
    string? icon
);

public sealed record CategoryResponse(
    Guid id,
    string name,
    CategoryType type,
    string color,
    string icon,
    bool archived
);
